beacon_node/network/src/sync/block_lookups/mod.rs

```diff
@@ -188,54 +201,93 @@ impl<T: BeaconChainTypes> BlockLookups<T> {
         cx: &mut SyncNetworkContext<T>,
     ) {
-        self.search_block_with(
+        let lookup = self.search_block_with(
             block_root,
             block,
             blobs,
             peer_source,
             LookupSource::UnknownParent,
         );
-        self.trigger_single_lookup(block_root, cx)
+
+        self.trigger_single_lookup(lookup, cx);
     }
 
+    /// Creates a lookup for the block with the given `block_root`, while caching other block
+    /// components we've already received. The block components are cached here because we haven't
+    /// imported it's parent and therefore can't fully validate it and store it in the data
+    /// availability cache.
+    ///
+    /// The request is not immediately triggered, and should be triggered by a call to
+    /// `trigger_lookup_by_root`.
     pub fn search_child_delayed(
         &mut self,
         block_root: Hash256,
         block: Option<Arc<SignedBeaconBlock<T::EthSpec>>>,
         blobs: Option<FixedBlobSidecarList<T::EthSpec>>,
         peer_source: &[PeerShouldHave],
     ) {
-        self.search_block_with(
+        let lookup = self.search_block_with(
             block_root,
             block,
             blobs,
             peer_source,
             LookupSource::UnknownParent,
         );
+
+        self.add_single_lookup(lookup)
     }
 
-    /// Attempts to trigger the request matching the given `block_root`. If the requests for
-    /// blocks and blobs could not have been made or are no longer required, this also removes the lookup.
-    pub fn trigger_single_lookup(&mut self, block_root: Hash256, cx: &mut SyncNetworkContext<T>) {
-        let mut drop_lookup = None;
-        for (index, single_block_request) in self.single_block_lookups.iter_mut().enumerate() {
-            if single_block_request
-                .block_request_state
-                .requested_block_root
-                == block_root
-                && !single_block_request.triggered
+    /// Attempts to trigger the request matching the given `block_root`.
+    pub fn trigger_single_lookup(
+        &mut self,
+        mut single_block_lookup: Option<SingleBlockLookup<SINGLE_BLOCK_LOOKUP_MAX_ATTEMPTS, T>>,
+        cx: &mut SyncNetworkContext<T>,
+    ) {
+        if let Some(single_block_lookup) = single_block_lookup.as_mut() {
+            if !single_block_lookup.triggered
+                && single_block_lookup.request_block_and_blobs(cx).is_ok()
             {
-                if single_block_request.request_block_and_blobs(cx).is_err() {
-                    drop_lookup = Some(index);
-                }
-                single_block_request.triggered = true;
+                single_block_lookup.triggered = true;
+            } else {
+                return;
             }
         }
-        if let Some(index) = drop_lookup {
-            self.single_block_lookups.swap_remove(index);
+        self.add_single_lookup(single_block_lookup)
```

Comment:
This would add the lookup even if it exists in the `single_block_lookups ` list right?

Reply:
The lookup should be `None` if it already exists due to the logic in `search_block_with`

Reply:
wait! in that case we'd drop duplicate lookups... damn that might explain the bug jimmy was seeing!